numba_cuda/numba/cuda/cudadrv/driver.py

```diff
@@ -2334,6 +2336,11 @@ def __int__(self):
         # The default stream's handle.value is 0, which gives `None`
         return self.handle.value or drvapi.CU_STREAM_DEFAULT
 
+    def __cuda_stream__(self):
+        if not self.handle.value:
+            return (0, drvapi.CU_STREAM_DEFAULT)
+        return (0, self.handle.value if USE_NV_BINDING else self.handle)
```

Comment:
This PR now full relies on having this protocol, even internally. But my concern is removing the internal fast paths we used to have could hurt the perf. I haven't caught up with the earlier discussions, not sure how much perf change would not impact & if this is acceptable.